### DISCUSSION TITLE
why is router.query only populated after second `useEffect` run?

### DISCUSSION THREAD

**Discussion by stefanprobst** [vercel/next.js#18268] [Help]
when accessing query params from `router.query` in a `useEffect` hook, they are only available after the second render pass, although `router.asPath` includes them.

Example:
```js
// pages/index.js
import { useEffect } from "react";
import { useRouter } from "next/router";

export default function HomePage() {
  const router = useRouter();
  useEffect(() => {
    console.log(router.asPath);
    console.log(router.query);
  }, [router.query]);
  return null;
}
```

on `http://localhost:3000/?hello=world` this will log:
```
/?hello=world
Object {  }
/?hello=world
Object { hello: "world" }
```

i realize that according to docs, the `query` object will only be populated after hydration - but i had assumed that `useEffect` would only run *after* the client was hydrated?

**Comment by jamesmosier** [comment]
Hi @stefanprobst. I believe this response in this issue explains why: https://github.com/vercel/next.js/issues/16019#issuecomment-693712695

**Reply by stefanprobst** [reply]
not sure it does? this comment talks about why `render.query` is not available in "server rendered code". i understand that. however, `useEffect` only executes client-side - and the query parameters clearly are available on `router.asPath` after the first render pass, but on `router.query` only after the second time. this is what is confusing to me,

**Reply by stefanprobst** [reply]
ok i think i got it now, thanks @jamesmosier for pointing me in the right direction. the confusing thing is that `router.query` and `new URL(router.asPath, baseUrl).searchParams` *don't* actually 100% represent the same thing.

**Reply by stefanprobst** [reply]
the docs actually have [this](https://nextjs.org/docs/routing/dynamic-routes#caveats):

```
Pages that are statically optimized by Automatic Static Optimization will be hydrated without their route parameters provided, i.e query will be an empty object ({}).

After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.
```

which means that in the initial `useEffect` run `router.query` will be an empty object. afaiu it will be updated in [`Container.componentDidMount`](https://github.com/vercel/next.js/blob/canary/packages/next/client/index.tsx#L166-L175) and is thus available in the second `useEffect` run.

the reason why `router.query` is initialized empty is to avoid hydration warnings. i.e. the following will *not* print a warning when visiting `http://localhost:3000/?hello=world`:

```js
import { useRouter } from "next/router";

export default function Home() {
  const router = useRouter();
  return <code>{JSON.stringify(router.query)}</code>;
}
```

**however**: the same does *not* hold for `router.asPath`. this will get [initialized with `window.location`](https://github.com/vercel/next.js/blob/canary/packages/next/client/index.tsx#L79). which means query parameters are already available by parsing `router.asPath` in the first `useEffect` run. it also means that using `router.asPath` outside of an `useEffect` will give a hydration warning. for example the following when visiting `http://localhost:3000/?hello=world`:

```js
import { useRouter } from "next/router";

export default function Home() {
  const router = useRouter();
  return <code>{router.asPath}</code>;
}
```
